Updater/Updater.php

```diff
@@ -44,9 +63,128 @@ public function update($referenceData, array $data, array $options = [])
         }
 
         foreach ($data as $propertyPath => $value) {
-            $this->propertyAccessor->setValue($referenceData, $propertyPath, $value);
+            $this->updateProperty($referenceData, $propertyPath, $value);
         }
 
-        return $this;
+        return $referenceData;
+    }
+
+    /**
+     * @param ReferenceDataInterface $referenceData
+     * @param string $propertyPath
+     * @param mixed $value
+     */
+    protected function updateProperty(ReferenceDataInterface $referenceData, $propertyPath, $value)
+    {
+        if ($this->propertyAccessor->isWritable($referenceData, $propertyPath)) {
+            if ($this->isAssociation($referenceData, $propertyPath)) {
+                $this->updateAssociatedEntity($referenceData, $propertyPath, $value);
+            } else {
+                $this->propertyAccessor->setValue($referenceData, $propertyPath, $value);
+            }
+        } elseif ($this->isAssociation($referenceData, 'translations')) {
+            $this->updateTranslation($referenceData, $propertyPath, $value);
+        }
+    }
+
+    /**
+     * Updates an entity linked to the reference data
+     *
+     * @param ReferenceDataInterface $referenceData
+     * @param string $propertyPath
+     * @param mixed $value
+     *
+     * @throws EntityNotFoundException
+     */
+    protected function updateAssociatedEntity(ReferenceDataInterface $referenceData, $propertyPath, $value)
+    {
+        $associationMapping = $this->getAssociationMapping($referenceData, 'translations');
+        $associationRepo  = $this->em->getRepository($associationMapping['targetEntity']);
+        $associatedEntity = $associationRepo->findOneBy(['code' => $value]);
```

Comment:
Memory leak and cascade detach is an integration concern and should be defined on the reference data doctrine mapping itself.
Btw, if I detach before persisting the entity, it cannot work as the uow will search to persist something it does not know anymore :)